File: C++/PalindromePartitioning.cc
// Given a string s, partition s such that every substring of the partition is a palindrome.

// Return all possible palindrome partitioning of s.

// For example, given s = "aab",
// Return

//   [
//     ["aa","b"],
//     ["a","a","b"]
//   ]

class Solution {
    bool isPalindrome(string& t) {
        int n = t.size() - 1;
        for (int i = 0; i < t.size() / 2; i++) {
            if (t[n - i] != t[i]) {
                return false;
            }
        }
        return true;
    }

    void collect(vector<vector<string>>& output, vector<string>& cur, string& s, int beg) {
        if (beg == s.size() && !cur.empty()) {
            output.push_back(cur);
        }
        for (int i = beg; i < s.size(); i++) {
            string t = s.substr(beg, i - beg + 1);
            if (isPalindrome(t)) {
                cur.push_back(t);
                collect(output, cur, s, i + 1);
                cur.pop_back();
            }
        }
    }
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> output;
        vector<string> cur;
        collect(output, cur, s, 0);
        return output;
    }
};

File: Java/PalindromePartitioning.java
// Given a string s, partition s such that every substring of the partition is a palindrome.

// Return all possible palindrome partitioning of s.

// For example, given s = "aab",
// Return

//   [
//     ["aa","b"],
//     ["a","a","b"]
//   ]

public class PalindromePartitioning {
    public List<List<String>> partition(String s) {
        List<List<String>> out = new ArrayList<>();
        partition(s, 0, new ArrayList<>(), out);
        return out;
    }

    void partition(String s, int i, List<String> cur, List<List<String>> out) {
        if (i == s.length()) {
            out.add(new ArrayList<>(cur));
            return;
        }

        for (int j = i+1; j <= s.length(); j++) {
            String t = s.substring(i, j);
            if (isPalindrome(t)) {
                cur.add(t);
                partition(s, j, cur, out);
                cur.remove(cur.size()-1);
            }
        }
    }

    boolean isPalindrome(String s) {
        int i = 0, j = s.length()-1;
        while (i < j) {
            if (s.charAt(i++) != s.charAt(j--))
                return false;
        }
        return true;
    }
}
